### PR TITLE
Test all functions for SD59x18

### PR DESCRIPTION
**Pull request by andreivladbrg**
Use foundry to test all the functions for SD59x18 using the inputs from the hardhat tests.

### CODE REVIEW

**Comment by paulrberg** [comment]
Thanks a lot for this.

- I know that technically it is cheaper to use `uncheckedAdd`, but we can just use `add` for the test values, to enhance readability. The gas reports are fudged by the test set arrays, anyway.
- I know that it's more verbose, but I would prefer to write `0.000000000000000001` instead of `1`.

**Comment by paulrberg** [comment]
Thanks a lot for this.

- All tests that revert should be refactored to start with `testCannot`.
- Great idea to name the functions in the test contracts like `minPermitted` and `maxPermitted` rather than using the minimum and maximum boundary values (e.g. `-59.794705707972522261`). But you should have also renamed the testing branches, for consistency `purposes.`
We should change the testing branches to match these names.
- I know that technically it is cheaper to use `uncheckedAdd`, but we can just use `add` for the test values, to enhance readability. The gas reports are fudged by the test set arrays, anyway.
- I know that it's more verbose, but I would prefer to write `0.000000000000000001` instead of `1`.

File: test/sd59x18/sqrt/sqrt.t.sol [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        sets.push(set({ x: MAX_SD59x18 }));
+        return sets;
+    }
+
+    function testSqrt__GreaterThanKeyThreshold() external parameterizedTest(greaterThanSets()) NotZero Positive {
+        vm.expectRevert(abi.encodeWithSelector(PRBMathSD59x18__SqrtOverflow.selector, s.x));
+        sqrt(s.x);
+    }
+
+    modifier LessThanKeyThreshold() {
```
Comment (paulrberg) [inline test/sd59x18/sqrt/sqrt.t.sol:55]: Let's rename this to `LessThanOrEqualToMaxPermitted`.

File: test/sd59x18/sqrt/sqrt.t.sol [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        _;
+    }
+
+    function lessThanSets() internal returns (Set[] memory) {
```
Comment (paulrberg) [inline test/sd59x18/sqrt/sqrt.t.sol:59]: Let's rename this to `lessThanOrEqualToMaxPermittedSets`.

File: test/sd59x18/sqrt/sqrt.t.sol [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        sets.push(set({ x: 1e18, expected: 1e18 }));
+        sets.push(set({ x: 2e18, expected: 1_414213562373095048 }));
+        sets.push(set({ x: E, expected: 1_648721270700128146 }));
+        sets.push(set({ x: 3e18, expected: 1_732050807568877293 }));
+        sets.push(set({ x: PI, expected: 1_772453850905516027 }));
+        sets.push(set({ x: 4e18, expected: 2e18 }));
+        sets.push(set({ x: 16e18, expected: 4e18 }));
+        sets.push(set({ x: 1e35, expected: 316227766_016837933199889354 }));
+        sets.push(set({ x: 1e36, expected: 1e27 }));
+        sets.push(set({ x: 12489131238983290393813.123784889921092801e18, expected: 111754781727_598977910452220959 }));
```
Comment (paulrberg) [inline test/sd59x18/sqrt/sqrt.t.sol:72]: Let's stick with the same format and redefine `x` as `12489131238983290393813_123784889921092801`.

File: test/sd59x18/sqrt/sqrt.t.sol [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        sets.push(
+            set({
+                x: 1889920002192904839344128288891377.732371920009212883e18,
```
Comment (paulrberg) [inline test/sd59x18/sqrt/sqrt.t.sol:75]: Let's stick with the same format and redefine `x` as `1889920002192904839344128288891377_732371920009212883`.

File: test/sd59x18/sqrt/sqrt.t.sol [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+                expected: 43473210166640613_973238162807779776
+            })
+        );
+        sets.push(set({ x: 1e58, expected: 1e38 }));
+        sets.push(set({ x: 5e58, expected: 223606797749978969640_917366873127623544 }));
+        sets.push(
+            set({
+                x: 57896044618658097711785492504343953926634.992332820282019728e18,
```
Comment (paulrberg) [inline test/sd59x18/sqrt/sqrt.t.sol:83]: Let's stick with the same format and redefine `x` as `57896044618658097711785492504343953926634_992332820282019728`.

File: test/sd59x18/log2/log2.t.sol [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        SD59x18 x = sd(-1);
+        vm.expectRevert(abi.encodeWithSelector(PRBMathSD59x18__LogInputTooSmall.selector, x));
+        log2(x);
+    }
+
+    modifier Positive() {
+        _;
+    }
+
+    function powerSets() internal returns (Set[] memory) {
```
Comment (paulrberg) [inline test/sd59x18/log2/log2.t.sol:37]: Let's rename this to `powerOfTwoSets`.

File: test/sd59x18/log2/log2.t.sol [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        sets.push(set({ x: 0.0625e18, expected: -4e18 }));
+        sets.push(set({ x: 0.125e18, expected: -3e18 }));
+        sets.push(set({ x: 0.25e18, expected: -2e18 }));
+        sets.push(set({ x: 0.5e18, expected: -1e18 }));
+        sets.push(set({ x: 1e18, expected: ZERO }));
+        sets.push(set({ x: 2e18, expected: 1e18 }));
+        sets.push(set({ x: 4e18, expected: 2e18 }));
+        sets.push(set({ x: 8e18, expected: 3e18 }));
+        sets.push(set({ x: 16e18, expected: 4e18 }));
+        sets.push(set({ x: 195e18, expected: 7_607330313749610674 }));
```
Comment (paulrberg) [inline test/sd59x18/log2/log2.t.sol:48]: This was a mistake in my JavaScript tests. 195 is not a power of two. This should have been `MAX_SD59x18.add(sd(1)).div(sd(2))` (the largest power of two which fits within `SD59x18`).

File: test/sd59x18/log2/log2.t.sol [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        return sets;
+    }
+
+    function testLog2__PowerOfTwo() external parameterizedTest(powerSets()) NotZero Positive {
+        SD59x18 actual = log2(s.x);
+        assertEq(actual, s.expected);
+    }
+
+    function notPowerSets() internal returns (Set[] memory) {
```
Comment (paulrberg) [inline test/sd59x18/log2/log2.t.sol:57]: Let's rename this to `notPowerOfTwoSets`.

File: test/sd59x18/ln/ln.t.sol [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        SD59x18 x = sd(-1);
+        vm.expectRevert(abi.encodeWithSelector(PRBMathSD59x18__LogInputTooSmall.selector, x));
+        ln(x);
+    }
+
+    modifier Positive() {
+        _;
+    }
+
+    function positiveSets() internal returns (Set[] memory) {
```
Comment (paulrberg) [inline test/sd59x18/ln/ln.t.sol:37]: Let's rename this to just `testSets`.

File: test/sd59x18/log10/log10.t.sol [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+        SD59x18 x = sd(-1);
+        vm.expectRevert(abi.encodeWithSelector(PRBMathSD59x18__LogInputTooSmall.selector, x));
+        log10(x);
+    }
+
+    modifier Positive() {
+        _;
+    }
+
+    function powerSets() internal returns (Set[] memory) {
```
Comment (paulrberg) [inline test/sd59x18/log10/log10.t.sol:37]: Let's rename this to `powerOfTenSets`.

File: test/sd59x18/log10/log10.t.sol [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+        sets.push(set({ x: 1e76, expected: 58e18 }));
+        return sets;
+    }
+
+    function testLog10__PowerOfTen() external parameterizedTest(powerSets()) NotZero Positive {
+        SD59x18 actual = log10(s.x);
+        assertEq(actual, s.expected);
+    }
+
+    function notPowerSets() internal returns (Set[] memory) {
```
Comment (paulrberg) [inline test/sd59x18/log10/log10.t.sol:62]: Let's rename this to `notPowerOfTenSets`.

File: test/sd59x18/gm/gm.t.sol [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+} from "~/SD59x18.sol";
+import { SD59x18__BaseTest } from "../SD59x18BaseTest.t.sol";
+
+contract SD59x18__GmTest is SD59x18__BaseTest {
+    // Biggest number whose square fits within int256
+    SD59x18 internal constant SQRT_MAX_SD59x18_DIV_BY_SCALE = SD59x18.wrap(240615969168004511545_033772477625056927);
+    // Smallest number whose square fits within int256
+    SD59x18 internal constant SQRT_MIN_SD59x18_DIV_BY_SCALE = SD59x18.wrap(-240615969168004511545_033772477625056927);
+
+    function zeroSets() internal returns (Set[] memory) {
```
Comment (paulrberg) [inline test/sd59x18/gm/gm.t.sol:25]: Let's rename this to `oneOperandZeroSets`.

File: test/sd59x18/gm/gm.t.sol [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+    function testGm__OneOperandZero() external parameterizedTest(zeroSets()) {
+        SD59x18 actual = gm(s.x, s.y);
+        assertEq(actual, s.expected);
+    }
+
+    modifier NotZeroOperands() {
+        _;
+    }
+
+    function negativeSets() internal returns (Set[] memory) {
```
Comment (paulrberg) [inline test/sd59x18/gm/gm.t.sol:41]: Let's rename this to `productNegativeSets`.

File: test/sd59x18/gm/gm.t.sol [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+        delete sets;
+        sets.push(set({ x: -7.1e18, y: 20.05e18, expected: ZERO }));
+        sets.push(set({ x: -1e18, y: PI, expected: ZERO }));
+        sets.push(set({ x: PI, y: -1e18, expected: ZERO }));
+        sets.push(set({ x: 7.1e18, y: -20.05e18, expected: ZERO }));
```
Comment (paulrberg) [inline test/sd59x18/gm/gm.t.sol:46]: No need to even pass the `expected` argument here. There is an overload of the `set` function which takes only `x` and `y` as arguments.

File: test/sd59x18/gm/gm.t.sol [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+    function testGm__ProductNegative() external parameterizedTest(negativeSets()) NotZeroOperands {
+        vm.expectRevert(abi.encodeWithSelector(PRBMathSD59x18__GmNegativeProduct.selector, s.x, s.y));
+        gm(s.x, s.y);
+    }
+
+    modifier PositiveProduct() {
+        _;
+    }
+
+    function overflowSets() internal returns (Set[] memory) {
```
Comment (paulrberg) [inline test/sd59x18/gm/gm.t.sol:59]: Let's rename this to `productOverflowSets`.

File: test/sd59x18/gm/gm.t.sol [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+        );
+        sets.push(
+            set({
+                x: SQRT_MAX_SD59x18_DIV_BY_SCALE.uncheckedAdd(sd(1)),
+                y: SQRT_MAX_SD59x18_DIV_BY_SCALE.uncheckedAdd(sd(1)),
+                expected: ZERO
+            })
+        );
+        sets.push(set({ x: MAX_WHOLE_SD59x18, y: 3, expected: ZERO }));
+        sets.push(set({ x: MAX_SD59x18, y: 2, expected: ZERO }));
```
Comment (paulrberg) [inline test/sd59x18/gm/gm.t.sol:78]: Again, no need to pass the `expected` argument here.

File: test/sd59x18/gm/gm.t.sol [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+        return sets;
+    }
+
+    function testGm__ProductOverflow() external parameterizedTest(overflowSets()) NotZeroOperands PositiveProduct {
+        vm.expectRevert(abi.encodeWithSelector(PRBMathSD59x18__GmOverflow.selector, s.x, s.y));
+        gm(s.x, s.y);
+    }
+
+    modifier NotOverflowProduct() {
```
Comment (paulrberg) [inline test/sd59x18/gm/gm.t.sol:87]: Let's rename this to `ProductNoOverflow`.

File: test/sd59x18/gm/gm.t.sol [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+        _;
+    }
+
+    function gmSets() internal returns (Set[] memory) {
```
Comment (paulrberg) [inline test/sd59x18/gm/gm.t.sol:91]: Let's rename this to `productNoOverflowSets`.

File: test/sd59x18/powu/powu.t.sol [hunk 19]
```diff
@@ -0,0 +1,215 @@
...
+
+    function testPowu__BaseAndExponentZero() external {
+        SD59x18 x = ZERO;
+        uint256 y = 0;
+        SD59x18 actual = powu(x, y);
+        SD59x18 expected = sd(1e18);
+        assertEq(actual, expected);
+    }
+
+    function baseZeroSets() internal returns (Set[] memory) {
```
Comment (paulrberg) [inline test/sd59x18/powu/powu.t.sol:33]: Let's rename this to `baseZeroExponentNotZeroSets`.

**Review by paulrberg — CHANGES_REQUESTED** [review]
Thanks a lot for this.

- All tests that revert should be refactored to start with `testCannot`.
- Great idea to name the functions in the test contracts like `minPermitted` and `maxPermitted` rather than using the minimum and maximum boundary values (e.g. `-59.794705707972522261`). But you should have also renamed the testing branches, for consistency purposes.
    - We should also change the testing branches to match these names.
- I know that technically it is cheaper to use `uncheckedAdd`, but we can just use `add` for the test values, to enhance readability. The gas reports are fudged by the test set arrays, anyway.
- I know that it's more verbose, but I would prefer to write `0.000000000000000001` instead of `1`.

File: test/sd59x18/div/div.t.sol [hunk 20]
```diff
@@ -0,0 +1,213 @@
...
+} from "~/SD59x18.sol";
+import { SD59x18__BaseTest } from "../SD59x18BaseTest.t.sol";
+
+contract SD59x18__DivTest is SD59x18__BaseTest {
+    SD59x18 internal constant MAX_SD59x18_DIV_BY_SCALE =
+        SD59x18.wrap(57896044618658097711785492504343953926634_992332820282019728);
+    SD59x18 internal constant MIN_SD59x18_DIV_BY_SCALE =
+        SD59x18.wrap(-57896044618658097711785492504343953926634_992332820282019728);
+
+    function testDiv__DenomitorZero() external {
```
Comment (paulrberg) [inline test/sd59x18/div/div.t.sol:27]: Typo: should be `Denominator`.

File: test/sd59x18/div/div.t.sol [hunk 21]
```diff
@@ -0,0 +1,213 @@
...
+        SD59x18 x = sd(1);
+        SD59x18 y = ZERO;
+        vm.expectRevert(stdError.divisionError);
+        div(x, y);
+    }
+
+    modifier DenominatorNotZero() {
```
Comment (paulrberg) [inline test/sd59x18/div/div.t.sol:34]: Typo: should be `Denominator`.

File: test/sd59x18/powu/powu.t.sol [hunk 22]
```diff
@@ -0,0 +1,215 @@
...
+        parameterizedTest(overflowOrUnderflowSets())
+        BaseNotZero
+        ExponentNotZero
+        ResultDoesNotOverflowOrUnderflowUint256
+    {
+        if (s.x.eq(SQRT_MIN_SD59x18.uncheckedSub(sd(1)))) {
+            vm.expectRevert(
+                abi.encodeWithSelector(
+                    PRBMathSD59x18__PowuOverflow.selector,
+                    57896044618658097711785492504343953926634992333271124942194_363391669063698216
```
Comment (paulrberg) [inline test/sd59x18/powu/powu.t.sol:110]: This value could have been set in the `expected` argument of the `set`.

File: test/sd59x18/powu/powu.t.sol [hunk 23]
```diff
@@ -0,0 +1,215 @@
...
+                )
+            );
+        } else {
+            vm.expectRevert(
+                abi.encodeWithSelector(PRBMathSD59x18__PowuOverflow.selector, MAX_SD59x18.uncheckedAdd(sd(1)))
```
Comment (paulrberg) [inline test/sd59x18/powu/powu.t.sol:115]: This value could have been set in the `expected` argument of the `set`.

File: test/sd59x18/powu/powu.t.sol [hunk 24]
```diff
@@ -0,0 +1,215 @@
...
+
+import {
+    E,
+    MAX_SD59x18,
+    MAX_WHOLE_SD59x18,
+    MIN_SD59x18,
+    MIN_WHOLE_SD59x18,
+    PI,
+    PRBMathSD59x18__Exp2InputTooBig,
+    PRBMathSD59x18__LogInputTooSmall,
```
Comment (paulrberg) [inline test/sd59x18/powu/powu.t.sol:12]: Unused imports.

We need a Solhint rule for this!

File: test/sd59x18/powu/powu.t.sol [hunk 25]
```diff
@@ -0,0 +1,215 @@
...
+    SD59x18,
+    ZERO,
+    powu
+} from "~/SD59x18.sol";
+import { PRBMath__MulDiv18Overflow } from "~/Helpers.sol";
+import { SD59x18__BaseTest } from "../SD59x18BaseTest.t.sol";
+
+contract SD59x18__PowuTest is SD59x18__BaseTest {
+    SD59x18 internal constant SQRT_MAX_SD59x18 = SD59x18.wrap(240615969168004511545033772477_625056927114980741);
+    SD59x18 internal constant SQRT_MIN_SD59x18 = SD59x18.wrap(-240615969168004511545033772477_625056927114980741);
```
Comment (paulrberg) [inline test/sd59x18/powu/powu.t.sol:23]: This should be renamed to `NEGATIVE_SQRT_MAX_SD59x18`.

`MIN_SD59x18`'s root is a complex number.

**Review by paulrberg — APPROVED** [review]
